pinot-controller/src/main/java/org/apache/pinot/controller/validation/RealtimeOffsetAutoResetManager.java

```diff
@@ -0,0 +1,240 @@
+ * under the License.
+ */
+package org.apache.pinot.controller.validation;
+
+import com.google.common.annotations.VisibleForTesting;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import org.apache.pinot.common.metrics.ControllerMetrics;
+import org.apache.pinot.controller.ControllerConf;
+import org.apache.pinot.controller.LeadControllerManager;
+import org.apache.pinot.controller.api.resources.Constants;
+import org.apache.pinot.controller.helix.core.PinotHelixResourceManager;
+import org.apache.pinot.controller.helix.core.periodictask.ControllerPeriodicTask;
+import org.apache.pinot.controller.helix.core.periodictask.RealtimeOffsetAutoResetHandler;
+import org.apache.pinot.controller.helix.core.realtime.PinotLLCRealtimeSegmentManager;
+import org.apache.pinot.spi.config.table.TableConfig;
+import org.apache.pinot.spi.stream.StreamConfig;
+import org.apache.pinot.spi.utils.IngestionConfigUtils;
+import org.apache.pinot.spi.utils.builder.TableNameBuilder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class RealtimeOffsetAutoResetManager extends ControllerPeriodicTask<RealtimeOffsetAutoResetManager.Context> {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(RealtimeOffsetAutoResetManager.class);
+  private final PinotLLCRealtimeSegmentManager _llcRealtimeSegmentManager;
+  private final PinotHelixResourceManager _pinotHelixResourceManager;
+  private final Map<String, RealtimeOffsetAutoResetHandler> _tableToHandler;
+  private final Map<String, Set<String>> _tableTopicsUnderBackfill;
+  private final Map<String, Set<String>> _tableEphemeralTopics;
+
+  public RealtimeOffsetAutoResetManager(ControllerConf config, PinotHelixResourceManager pinotHelixResourceManager,
+      LeadControllerManager leadControllerManager, PinotLLCRealtimeSegmentManager llcRealtimeSegmentManager,
+      ControllerMetrics controllerMetrics) {
+    super("RealtimeOffsetAutoResetManager", config.getRealtimeOffsetAutoResetFrequencyInSeconds(),
+        config.getRealtimeOffsetAutoResetInitialDelaySeconds(), pinotHelixResourceManager,
+        leadControllerManager, controllerMetrics);
+    _llcRealtimeSegmentManager = llcRealtimeSegmentManager;
+    _pinotHelixResourceManager = pinotHelixResourceManager;
+    _tableToHandler = new ConcurrentHashMap<>();
+    _tableTopicsUnderBackfill = new ConcurrentHashMap<>();
+    _tableEphemeralTopics = new ConcurrentHashMap<>();
+  }
+
+  @Override
+  protected RealtimeOffsetAutoResetManager.Context preprocess(Properties periodicTaskProperties) {
+    RealtimeOffsetAutoResetManager.Context context = new RealtimeOffsetAutoResetManager.Context();
+    // Fill offset back fill job required info
+    if (periodicTaskProperties.containsKey(context._backfillJobPropertyKeys.toArray()[0])) {
```

Comment:
can you add pointers for a test or a comment/javaDoc for an example of periodTaskProperties.